src/elements/StoryBookmark/StoryBookmark.tsx

```diff
@@ -0,0 +1,78 @@
+import { StoryBookmarkNode, STORY_BOOKMARK_NODE_TYPE } from '@prezly/slate-types';
+import { StoryBookmarkLayout } from '@prezly/slate-types';
+import React, { useRef, useState, useMemo } from 'react';
+
+import { useResizeObserver, utils } from '../../lib';
+
+import { BookmarkCard } from '../../components';
+import { useElementContext } from '../../RendererContext';
+
+interface StoryBookmarkBlockProps {
+    node: StoryBookmarkNode;
+}
+
+const HORIZONTAL_LAYOUT_MIN_WIDTH = 480;
+
+export function StoryBookmark({ node }: StoryBookmarkBlockProps) {
+    const ctx = useElementContext(STORY_BOOKMARK_NODE_TYPE);
+    const card = useRef<HTMLDivElement | null>(null);
+    const [isSmallViewport, setSmallViewport] = useState(false);
+
+    useResizeObserver(card.current, function (entries) {
+        entries.forEach(function (entry) {
+            setSmallViewport(entry.contentRect.width < HORIZONTAL_LAYOUT_MIN_WIDTH);
+        });
+    });
+
+    if (!ctx) {
+        return null;
+    }
+
+    const story = ctx.getStory(node.story.uuid);
+
+    const showThumbnail = node.show_thumbnail && story.oembed.thumbnail_url;
+
+    const isEmpty =
+        !showThumbnail &&
+        utils.isEmptyText(story.oembed.title) &&
+        utils.isEmptyText(story.oembed.description);
+
+    const actualLayout = useMemo(() => {
+        if (!showThumbnail) {
+            return StoryBookmarkLayout.HORIZONTAL;
+        } else if (isSmallViewport) {
+            return StoryBookmarkLayout.VERTICAL;
+        } else {
+            return node.layout;
+        }
+    }, [showThumbnail, isSmallViewport, node.layout]);
```

Comment:
I am sure there is no performance gain from wrapping `actualLayout` calculation into `useMemo()`. Most likely, the opposite: `useMemo()` will make it slower for no reason.